src/RegistryCompilerPass.php

```diff
@@ -58,23 +73,35 @@ private function buildRegistry(ContainerBuilder $container, $registryId, array $
 					}
 				}
 
-				$collection->add($id, $options);
+				$collection->add([$id, $optionsResolver->getMetadata($options)], $options);
 			}
 
 		}
 
 		$targetId = $registryId;
 		$methodName = $tagOptions['method'];
+		$useMetadata = null !== $tagOptions['metadata'];
+
 		if ($tagOptions['use_collection']) {
 			$targetId = "{$registryId}.collection";
-			$container->setDefinition($targetId, new Definition(\ArrayObject::class))->setPublic(false);
+
+			$collectionClass = $useMetadata ? MetaDataIterator::class : \ArrayObject::class;
+			$container->setDefinition($targetId, new Definition($collectionClass))->setPublic(false);
```

Comment:
`setFoobars([])` vs `addFoobar($)` is a performance boost for you? I’m leaning towards removing the `use_collection`, since it’s trivial to do this part manually if needed.

I could switch it to `use_adder` vs `use_setter`… But then I’d have to throw exception on `use_setter` with `metadata`.